core/src/main/java/org/infinispan/persistence/spi/CacheWriter.java

```diff
@@ -35,4 +39,30 @@
     * @throws PersistenceException in case of an error, e.g. communicating with the external storage
     */
    boolean delete(Object key);
+
+   /**
+    * Persist all provided entries to the store in a single batch update. If this is not supported by the
+    * underlying store, then entries are written to the store individually via {@link #write(MarshalledEntry)}.
+    * As the execution order of MarshalledEntries is not guaranteed, you should ensure that only a single entry exists
+    * in the Collection for a given key.
+    *
+    * @param entries a Collection of MarshalledEntry to be written to the store.
+    * @throws NullPointerException if entries is null.
+    */
+   default void writeBatch(Collection<MarshalledEntry<? extends K, ? extends V>> entries) {
```

Comment:
I'm still not sure how the batching in `OffHeapDataContainer.entrySet()` was supposed to work... the consumer of the stream can't say "give me the elements from the next mapper invocation", it will still read elements from an iterator one by one. At most you could implement a spliterator whose `trySplit()` method uses the lock/address/pointer boundaries, but I doubt the default `flatMap()` will do that even if the [JDK-8075939](https://bugs.openjdk.java.net/browse/JDK-8075939) is fixed.

BTW, even though there isn't a `flatMapToObj` method, you could use `intStream.mapToObj(i -> stream).flatMap(stream -> stream)`.